src/machi_lifecycle_mgr.erl

```diff
@@ -0,0 +1,1079 @@
+
+make_ast_run_env() ->
+    {_KV_old=gb_trees:empty(), _KV_new=gb_trees:empty(), _IsNew=false}.
+
+env_host_exists(Name, E) ->
+    Key = {kv,{host,Name}},
+    case d_find(Key, E) of
+        error ->
+            false;
+        {ok, _} ->
+            true
+    end.
+
+env_flu_exists(Name, E) ->
+    Key = {kv,{flu,Name}},
+    case d_find(Key, E) of
+        error ->
+            false;
+        {ok, _} ->
+            true
+    end.
+
+get_host_client_interface(HostName, E) ->
+    Key = {kv,{host,HostName}},
+    case d_find(Key, E) of
+        error ->
+            false;
+        {ok, {host, _Name, _AdminI, ClientI, _Props}} ->
+            {ok, ClientI}
+    end.
+
+host_port_is_assigned(HostName, Port, {KV_old, KV_new, _}) ->
+    L = gb_trees:to_list(KV_old) ++ gb_trees:to_list(KV_new),
+    FLU_Ts = [V || {{kv,{flu,_}}, V} <- L],
+    case [V || {flu, _Nm, Host_, Port_, _Ps}=V <- FLU_Ts,
+               Host_ == HostName, Port_ == Port] of
+        [{flu, Name, _Host, _Port, _Ps}] ->
+            {true, Name};
+        [] ->
+            false
+    end.
+
+d_find(Key, {KV_old, KV_new, IsNew}) ->
+    %% Bah, use 'dict' return value convention.
+    case gb_trees:lookup(Key, KV_new) of
+        {value, Val} when IsNew ->
+            {ok, Val};
+        _ ->
+            case gb_trees:lookup(Key, KV_old) of
+                {value, Val} ->
+                    {ok, Val};
+                _ ->
+                    error
+            end
+    end.
+
+d_get(Key, {KV_old, KV_new, IsNew}) ->
+    %% Bah, use 'dict' return value convention.
+    %% Assume key exists, fail if not found.
+    case gb_trees:lookup(Key, KV_new) of
+        {value, Val} when IsNew ->
+            Val;
+        _ ->
+            case gb_trees:lookup(Key, KV_old) of
+                {value, Val} ->
+                    Val
+            end
+    end.
+
+d_store(Key, Val, {KV_old, KV_new, false}) ->
+    {gb_trees:enter(Key, Val, KV_old), KV_new, false};
+d_store(Key, Val, {KV_old, KV_new, true}) ->
+    {KV_old, gb_trees:enter(Key, Val, KV_new), true}.
+
+d_erase({tmp,_}=Key, {KV_old, KV_new, IsNew}) ->
+    {gb_trees:delete_any(Key, KV_old), gb_trees:delete_any(Key, KV_new), IsNew}.
+
+switch_env_dict({KV_old, KV_new, false}) ->
+    {KV_old, KV_new, true};
+switch_env_dict({_, _, true}) ->
+    A = switch_old_and_new,
+    err("Duplicate ~p", [A], A).
+
+n(L) ->
+    lists:sort(L).
+
+err(Fmt, Args, AST) ->
+    throw({str(Fmt, Args), AST}).
+
+str(Fmt, Args) ->
+    lists:flatten(io_lib:format(Fmt, Args)).
+
+%% We won't allow 'atom' style proplist members: too difficult to normalize.
+%% Also, no duplicates, again because normalizing useful for checksums but
+%% bad for order-based traditional proplists (first key wins).
+
+is_proplisty(Props) ->
+    is_list(Props) andalso
+    lists:all(fun({_,_})             -> true;
+                 %% nope: (X) when is_atom(X) -> true;
+                 (_)                 -> false
```

Comment:
It's a consistency pattern, I think, though not explicitly mentioned, sorry.  I believe (though not very strongly) that the proplist should pass through as-is or be rejected.